proposals/content-script-top-frame-matching.md

```diff
@@ -0,0 +1,161 @@
+
+```json5
+{
+  // ... existing content_script properties like "matches", "exclude_matches", etc.
+
+  // If specified: Only inject if the top-level frame's origin matches at least one of these patterns.
+  "top_frame_matches": ["MatchPattern"],
+
+  // If specified: Only inject if the top-level frame's origin isn't a match for any pattern.
+  "exclude_top_frame_matches": ["MatchPattern"]
+}
+```
+
+*Where `MatchPattern` is a string conforming to the standard [Extension Match Pattern syntax](https://developer.chrome.com/docs/extensions/develop/concepts/match-patterns). 
+
+
+#### `scripting.RegisteredContentScript` Type
+
+The `RegisteredContentScript` type used by `scripting.registerContentScripts()` and `scripting.updateContentScripts()` is expanded similarly:
+
+```typescript
+dictionary RegisteredContentScript {
+  // ... existing RegisteredContentScript properties like "matches", "excludeMatches", etc.
+  
+  // If provided, only inject if the top-level frame's origin matches at least one of these patterns.
+  MatchPattern[]? topFrameMatches;
+
+  // If provided, Only inject if the top-level frame's origin isn't a match for any pattern.
+  MatchPattern[]? excludeTopFrameMatches;
+}
+```
+
+
+### Behavior / Implementation
+
+1.  **Validation:** When processing `content_scripts` from `manifest.json` or a call to `scripting.registerContentScripts` / `scripting.updateContentScripts`:
+    *   The browser must first validate all patterns provided in `topFrameMatches` and `excludeTopFrameMatches` as they would validate patterns provided by `matches` and `excludeMatches`. 
+    *   Additionally, if any pattern contains a path component other than the wildcard path `/*` (i.e., it specifies a specific path like `/foo` or `/bar/*`), the browser must treat this as an error. Patterns without an explicit path or those explicitly using `/*` are considered valid. This restriction ensures these patterns are intended to match origins.
+        *   For static declarations in `manifest.json`, this should result in a manifest parsing error, preventing the extension from loading.
+        *   For dynamic API calls (`registerContentScripts`, `updateContentScripts`), the promise must be rejected with an appropriate error (e.g., `Match patterns for top_frame_matches/exclude_top_frame_matches must not specify a path.`).
+
+2.  **Injection Logic:** Assuming validation passes, a content script will be injected into a frame if and only if *all* the following conditions are met:
+    *   All existing checks based on the frame's own URL and context are satisfied (e.g., `matches`, `excludeMatches`).
+    *   And if `topFrameMatches` was specified, the **top-level document's origin** matches at least one pattern in `topFrameMatches`.
+    *   And if `excludeTopFrameMatches` was specified, the **top-level document's origin** does *not* match any pattern in `excludeTopFrameMatches`.
+
+
```

Comment:
```suggestion
If any of the match patterns is not supported or understood by the browser, a soft error should be thrown. In `topFrameMatches`, the pattern will be skipped. If `topFrameMatches` is, or thus will end up being empty, no content scripts will be injected.

If `excludeTopFrameMatches` is empty, the property will be ignored and not considered when injecting content scripts.
```

We should still determine what should happen in case invalid patterns show up in excludeTopFrameMatches.